components/brave_rewards/core/wallet_provider/solana/solana_wallet_provider.cc

```diff
@@ -0,0 +1,198 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_rewards/core/wallet_provider/solana/solana_wallet_provider.h"
+
+#include <cmath>
+#include <utility>
+#include <vector>
+
+#include "base/base64url.h"
+#include "base/strings/strcat.h"
+#include "base/strings/string_util.h"
+#include "brave/components/brave_rewards/core/common/signer.h"
+#include "brave/components/brave_rewards/core/endpoint/rewards/rewards_util.h"
+#include "brave/components/brave_rewards/core/global_constants.h"
+#include "brave/components/brave_rewards/core/logging/event_log_keys.h"
+#include "brave/components/brave_rewards/core/state/state_keys.h"
+#include "brave/components/brave_rewards/core/wallet/wallet_util.h"
+#include "brave/components/brave_rewards/core/wallet_provider/linkage_checker.h"
+#include "net/base/url_util.h"
+#include "url/gurl.h"
+
+namespace brave_rewards::internal {
+
+namespace {
+
+constexpr char kSplBatTokenMint[] =
+    "EPeUFDgHRxs9xxEPVaL6kfGQvCon7jmAWKVUHuux1Tpz";
+constexpr base::TimeDelta kPollingInterval = base::Seconds(10);
+constexpr base::TimeDelta kPollingTimeout = base::Minutes(5);
+
+std::string UsernameFromAddress(const std::string& address) {
+  size_t length = address.length();
+  if (length < 10) {
+    return address;
+  }
+  std::string_view address_view(address);
+  return base::StrCat({address_view.substr(0, 5), "...",
+                       address_view.substr(length - 4, length)});
+}
+
+}  // namespace
+
+SolanaWalletProvider::SolanaWalletProvider(RewardsEngineImpl& engine)
+    : RewardsEngineHelper(engine), WalletProvider(engine) {}
+
+SolanaWalletProvider::~SolanaWalletProvider() = default;
+
+const char* SolanaWalletProvider::WalletType() const {
+  return constant::kWalletSolana;
+}
+
+void SolanaWalletProvider::AssignWalletLinks(
+    mojom::ExternalWallet& external_wallet) {
+  std::string explorer_url =
```

Comment:
nit: I believe we'd want to use a URL builder here rather than concatenating a string. would you mind changing this?

Reply:
Done